CIPs/cip-0032.md

```diff
@@ -0,0 +1,117 @@
+
+We will introduce a new `AttestationSlasher` contract which allows slashing of stake if a completed attestation hasn't been observed in the past `B` blocks. To achieve this, we will upgrade the `Attestations` contract to register the selection and completion events in the new `AttestationSlasher` contract. The `AttestationSlasher` will contain the following state to track the latest block number that an attestation was requested/completed on:
+
+```js
+// Tracks the last time an attestation was requested from an issuer (AS node)
+mapping(address => uint256) public lastAttestationRequested;
+// Tracks the last time an attestation was completed by an issuer
+mapping(address => uint256) public lastAttestationCompleted;
+// Tracks the last time an issuer was last slashed
+mapping(address => uint256) public lastSlashed;
+```
+
+A slashing event may occur when the slashing initiator calls `slash()` with the slashable address. Before slashing, the slashing contract checks the following:
+
+1. The validator has been elected continuously for the past `B` blocks.
+2. The issuer was selected for at least one attestation in the past `B` blocks.
+3. The validator wasn't slashed by the `AttestationSlasher` within the past `B` blocks.
+4. The address hasn't completed an attestation in the past `B` blocks.
+5. The addres hasn't submitted `SignalOnline` event in the last `S` blocks. (see [below](#signal-online)).
+
+If the validator was intermittently elected, the contract will only consider the past `B` blocks in which the validator was *continuously* elected. For example, if a validator with a downed attestation node was elected for .5 `B` blocks, was unelected, and re-elected for .5 `B` blocks, then it will not be eligible for slashing since the slasher restarts counting when the validator was re-elected. Additionally, if the validator was elected but was never selected to provide an attestation in the entire period of `B` blocks, then they are not eligible for slashing.
+
+### Signal online
+
+In the unlikely event that a validator was never selected for an attestation during the period of `B` blocks, it's possible for a malicious actor to request multiple attestations with the goal of randomly getting paired with this validator and intentionally leaving the attestation incomplete. To prevent this, we're building into the AS the ability to signal to the `AttestationSlasher` that the service is still online if a slashing event becomes close by signing and submitting a transaction to the slasher contract. This requires that the AS signer key to be minimally funded to pay the gas fees of the transaction, however this adds more piece of mind for validators to be sure they won't be incorrectly slashed. Additionally, we'll be adding two new Prometheus events to the AS to alert a validator when the slashing period is nearly reached and another event to alert when the AS signer key's balance is low.
```

Comment:
```suggestion
In the unlikely event that a validator was never selected for an attestation during the period of `B` blocks, it's possible for a malicious actor to request multiple attestations with the goal of randomly getting paired with this validator and intentionally leaving the attestation incomplete. To prevent this, we're building into the AS the ability to signal to the `AttestationSlasher` that the service is still online if a slashing event becomes close by signing and submitting a transaction to the slasher contract. This requires that the AS signer key to be minimally funded to pay the gas fees of the transaction, however this adds more peace of mind for validators to be sure they won't be incorrectly slashed. Additionally, we'll be adding two new Prometheus events to the AS to alert a validator when the slashing period is nearly reached and another event to alert when the AS signer key's balance is low.
```